iptables/src/iptables_ext.cpp

```diff
@@ -0,0 +1,216 @@
+/**
+ * Copyright (c) 2018 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *  Parts of this file are also:
+ *
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under both the Apache 2.0 license (found in the
+ *  LICENSE file in the root directory of this source tree) and the GPLv2 (found
+ *  in the COPYING file in the root directory of this source tree).
+ *  You may select, at your option, one of the above-listed licenses.
+ */
+
+#include <osquery/sdk.h>
+
+#include <arpa/inet.h>
+#include <netdb.h>
+#include <sstream>
+
+#include <boost/algorithm/string/trim.hpp>
+
+#include <trailofbits/extutils.h>
+
+#include "iptables_ext.h"
+#include "utils.h"
+
+using namespace osquery;
+
+namespace trailofbits {
+osquery::QueryData IptablesExtTable::generate(osquery::QueryContext& context) {
+  osquery::QueryData results;
+
+  MatchMap match_map;
+  auto s = parseIptablesSave(match_map);
+
+  if (!s.ok()) {
+    TLOG << "Error fetching matches from iptables-save: " << s.toString();
+  }
+
+  for (auto& table : getIptablesNames()) {
+    const auto& matches = match_map.find(table);
+
+    if (matches == match_map.end()) {
+      TLOG << "Couldn't associate table " << table << " with a list of matches";
+      return results;
+    }
+
+    s = genIptablesRules(table, matches->second, results);
+
+    if (!s.ok()) {
+      TLOG << "Error while fetching table rules: " << s.toString();
+      return results;
+    }
+  }
+
+  return results;
+}
+
+osquery::Status IptablesExtTable::genIptablesRules(
+    const std::string& filter,
+    const MatchChain& matches,
+    osquery::QueryData& results) {
+  // Initialize the access to iptc
+  auto handle = iptc_init(filter.c_str());
+  if (handle == nullptr) {
+    return osquery::Status(1, "Couldn't initialize iptables handle");
+  }
+
+  // Iterate through chains
+  for (auto chain = iptc_first_chain(handle); chain != nullptr;
+       chain = iptc_next_chain(handle)) {
+    // NOTE(ww): Rules are 1-based in libiptc, as evidenced by
+    // iptc_read_counter.
+    unsigned long ruleno = 1;
+    const auto& match_list = matches.find(chain);
```

Comment:
small nit, it could help readability to create a local variable for `match_list->second` which is _actually_ the `MatchList` (i spent a bit of time confused why a `std::vector` would have a `->second`)

Reply:
Yup -- `match_list` was the original variable (and was indeed a `std::vector`) before I introduced some more field extraction. I'll make it `match_pair` and add a new `match_list` with the expected type.